llvm/lib/Object/COFFObjectFile.cpp

```diff
@@ -800,6 +800,219 @@ Error COFFObjectFile::initLoadConfigPtr() {
+                                         ") is too small");
+          if (Header->BlockSize % sizeof(uint32_t))
+            return createStringError(
+                object_error::parse_failed,
+                "Unaligned ARM64X relocations block size (" +
+                    Twine(Header->BlockSize) + ")");
+          if (Header->BlockSize > RelocContents.size())
+            return createStringError(object_error::parse_failed,
+                                     "ARM64X relocations block size (" +
+                                         Twine(Header->BlockSize) +
+                                         ") is too large");
+          if (Header->PageRVA & 0xfff)
+            return createStringError(object_error::parse_failed,
+                                     "Unaligned ARM64X relocations page RVA (" +
+                                         Twine(Header->PageRVA) + ")");
+
+          ArrayRef<uint16_t> Relocs(
+              reinterpret_cast<const uint16_t *>(RelocContents.data() +
+                                                 sizeof(*Header)),
+              (Header->BlockSize - sizeof(*Header)) / sizeof(uint16_t));
+          RelocContents = RelocContents.drop_front(Header->BlockSize);
+
+          while (!Relocs.empty()) {
+            if (!Relocs[0]) {
+              if (Relocs.size() != 1)
+                return createStringError(
+                    object_error::parse_failed,
+                    "Unexpected ARM64X relocations terminator");
+              break;
+            }
+
+            uint16_t Arg = Relocs[0] >> 14, RelocSize = 1, Size;
+            switch ((Relocs[0] >> 12) & 3) {
+            case COFF::IMAGE_DVRT_ARM64X_FIXUP_TYPE_ZEROFILL:
+              Size = 1 << Arg;
+              break;
+            case COFF::IMAGE_DVRT_ARM64X_FIXUP_TYPE_VALUE:
+              if (!Arg)
+                return createStringError(
+                    object_error::parse_failed,
+                    "Invalid ARM64X relocation value size (0)");
+              Size = 1 << Arg;
+              RelocSize += Size / sizeof(Relocs[0]);
+              break;
+            case COFF::IMAGE_DVRT_ARM64X_FIXUP_TYPE_DELTA:
+              ++RelocSize;
+              Size = sizeof(uint32_t);
+              break;
+            default:
+              return createStringError(object_error::parse_failed,
+                                       "Invalid relocation type");
+            }
+            if (Header->PageRVA) {
+              uint64_t IntPtr;
+              uint16_t Offset = Relocs[0] & 0xfff;
+              if (Offset % Size)
+                return createStringError(object_error::parse_failed,
+                                         "Unaligned ARM64X relocation RVA (" +
+                                             Twine(Header->PageRVA + Offset) +
+                                             ")");
+              if (Error E = getRvaPtr(Header->PageRVA + Offset + Size, IntPtr,
```

Comment:
It's not a leftover, it's for error handling. This makes sure that relocation RVAs can be trusted and there is a test for that, see `ERR-RELOC-RVA`.

Yes, iterators implement the same logic, but having it here assumes that they can trust the input they operate on (where otherwise we wouldn't have a way to propagate errors). For that, we need to enumerate relocs here too. Looking at it again, I noticed that I missed one more check, to make sure that there is no reloc entry with a size exceeding the block size, I will add it.

We could probably skip that particular check from iterators point of view: they don't really care about reloc RVA, but it still seems right to guarantee callers that we return reasonable data.

Reply:
Does something like https://github.com/cjacek/llvm-project/commit/e60a959f759426dfcfcffb0ef978465cd8f4a899 look better?

I can use `Arm64XRelocRef` here to abstract them, but I'm not sure it makes it easier to follow. It saves a few bits operations, but the validation code needs to know the layout anyway.

I also updated a draft of a followup: https://github.com/cjacek/llvm-project/commit/299b6cb3b0c88f0cc5b29e8768d87246239d1aa5

Reply:
Yes, that commit makes it a little bit more palatable.

I guess part of this comes from the overall design of these classes; that we normally do all checking upfront, and then we just return data that is supposed to be known safe. For this kind of data structure, which is variable length in two layers, the verification requires doing the whole iteration first once upfront, and then later via the iterators. If it would be possible to return such validation errors from the iterators, instead of upfront, I feel we would be able to get rid of the duplication entirely.

Then again - I'm not sure what others, with more experience with our `ObjectFile` classes, think about this - @efriedma-quic, any opinions?

Alternatively, would it be possible to collapse it even further, into basically what the caller would do, e.g. something like this:
```
    for (auto reloc : dynamic_relocs()) {
      switch (reloc.getType()) {
      case COFF::IMAGE_DYNAMIC_RELOCATION_ARM64X: {
        for (auto Arm64XReloc : reloc.arm64x_relocs())
          getRvaPtr(Arm64XReloc.getRVA());
```

The iterator interfaces themselves don't expose many opportunities for validation, but would it be possible to add e.g. a `Error validate();` method on the iterators, which would encapsulate all this (checking validity of the RVAs, checking for alignment issues etc), so that the check could boil down to just this:
```
    for (auto reloc : dynamic_relocs())
        if (Error e = reloc.validate())
            return e;
```

As I'm a little unsure whether I'm leading you in the right direction or just diverging from the common LLVM ObjectFile code style, I'd wait to see if @efriedma-quic agrees or disagrees with my suggestions here.

Reply:
> Alternatively, would it be possible to collapse it even further, into basically what the caller would do, e.g. something like this:

No, it wouldn't be enough. We need to validate that the content is properly terminated to make `moveNext` safe.

Reply:
I think I can see what you mean now. `Arm64XRelocRef` class does not have enough info to do proper range checks on its own, but we may just pass `COFFObjectFile` to `validate()`. With the new scheme, I think we may even avoid explicit `moveNext` call (we may have an invalid header pointer in the iterator, but it should be fine as long as we validate it before actually using it).

I pushed a new version, thanks for the feedback!

Reply:
Thanks! Yes, this is exactly what I meant. The total amount of code perhaps isn't quite as small, and unfortunately, some of the terminator checking bits probably are less straightforward than before, but we don't duplicate the whole layout/iteration handling logic - which was what I disliked before.